core/src/main/scala/scalaz/MonadListen.scala

```diff
@@ -11,4 +11,26 @@ trait MonadListen[F[_], W] extends MonadTell[F, W] {
 
 object MonadListen {
   def apply[F[_], W](implicit ML: MonadListen[F, W]) = ML
+
+  /** The Free instruction set for MonadListen */
+  sealed abstract class Ast[S, A]
+  final case class Listen[F[_], S, A](ma: Free[F, A]) extends Ast[S, (A, S)]
```

Comment:
We shouldn't have any self-recursive algebras.

Reply:
1. why not?
2. what do you suggest instead?

Reply:
also what do you mean by self recursive? Is that more general than GADT?